### ISSUE TITLE
Block mechanic

### ISSUE THREAD

**Issue by Turtyo** [Saplings-Projects/1M_sub#111]
## Feature Description

Implement the mechanic that allows a player to gain shield to block incoming enemy damage.

## Motivation

Since health is a resource that's difficult to come by, being able to block incoming damage every turn is important.

## Concept art / Designs

Show somewhere how much damage will be blocked.

## Acceptance Criteria

[Define the criteria that need to be met for this feature to be considered complete.]

- Enemy damage starts to deplete the block first, then the health
- Block is reset to 0 each turn
- If the damage is more than the block, the excess damage carries over to damage to health (if there is 5 block remaining and 7 damage, we take 2 health damage)

## Proposed Solution

We can probably copy the health component to do a block component, with a reset to 0 in a `on_turn_start` function.

**Comment by JonaLam** [comment]
A lot of the damage effects in the game tend to use the 'modify_health' function, so we need to either change that or have something intercept it to make armor work. I think the best solution would be to either change the health component directly to have armot or add an armor component which the healthcomponent has a direct refrence to, and then do the calculations in the 'modefy_health' function.

As for the display, I think just a simple shield with a number would work fine, at least as a placeholder

**Comment by Turtyo** [comment]
Actually, all the damage effects use `modify_health`.
I think having a block value in the health component and checking that when applying the logic of modify health would work fine; I think an even better solution is to leave `modify_health` unchanged (leave it to just do what's it meant to do).

Instead make a new function `modify_health_and_block`, which decides if it should modify health or modify block, then make a `modify_block` function similar to `modify_health`. Also need to add an `on_turn_start` to reset the block to 0.

I thought about a new class. From an organization standpoint, it would be better to have the block as a separate component; the problem I have with it is that we can't have reference to other component inside a given component (because then you lose the advantage of components, ie it being modular). So I think in the end having the block with the health still makes sense.

A simple shield with a number will do it for now, we can modify the asset later.

**Comment by JonaLam** [comment]
I assume the function to be made should be in the health component, and in that case, maybe it should be called something more generic like `take_damage` as it will probably be what the damage functions call when they are going to do their thing.

 It would be possible to put the block component in it's own script which the health component always has a refrence to, so that we prevent the health component script from being too big, but that might be unnecesary complexity.

**Comment by Turtyo** [comment]
`modify_health` makes more sense because it's not just for taking damage, but also to heal. Healing with a function called `take_damage` would be a bit strange.

I think that would be unnecessary complexity for two reasons:
- having a component structure kind of implies you don't assume the existence of one component for other components (at least that's how I view it)
- I don't see a situation where an entity would have health  but no block, so separating them feels unnecessary

**Comment by JonaLam** [comment]
The `take_damage` function would be instead of a new function `modify_health_and_block`, not `modify health`. the reason for the name is that I think that the name `modify_health_and_block` would imply that it could apply health and block, which I don't think it should as I can't think of any way to do that which won't be confusing. `take_damage` already implies removing armor first, though it could be confusing if the user doesn't have that assumtion so there might be better names

For giving block it should probably be just a simple function like `add_block` or `apply_block` as that can't be misunderstood.

**Comment by Turtyo** [comment]
The user doesn't see the function names :laughing: 
i'm getting a bit confused, could you make a little summary of what names you would make, what those functions do and eventually which function they call ?

**Comment by JonaLam** [comment]
> The user doesn't see the function names 😆
Ye I meant if someone else were gonna use the functions

I'll summuraize my proposed solution since it's getting a bit confusing

The health component wil be extended to have block, which wil be held in an int. There should be three helper functions `add_block`, `remove_block`, and `set_block`, as well as maybe `reset_block` which would put block to 0, which might be needed to work with signals. To make blocking work with the damageing system I think making a new function called `take_damage` which would do the logic of retracting the block before the health and such, all of the damage effects which use `modify_health` wil be changed to use the new function. The new `take_damage` function wil still use `modify_health` if the target loses health as well as the new `remove_block` function to do that.

**Comment by Turtyo** [comment]
I would do something like that instead:
- keep `modify_health` signature as is (since changing this makes a lot of changes in other places, better keep the API)
- add a `modify_block`, which works similarly to `modify_health`, but on block

Inside `modify_health`, if we are taking damage, call `modify_block` (not in the case were we are healing, we don't want to block the healing).
`modify_block` returns an int, which is the amount of damage to be dealt to the health component after calculating the block. There are two cases:
- the block amount was bigger than the damage amount, the block is reduced and `modify_block` returns 0 (we absorbed all the damage)
- the block amount was lower than the damage amount, the block is reduced to 0 and we return a number bigger than 0

In both cases, we can just replace the damage that would be dealt to the health by the number returned by `modify_block`.
`modify_block` can also be called by effects that add block or remove block

I think a `reset_block` function would be good, purely for clarity. What do you think of this @JonaLam ?

**Comment by JonaLam** [comment]
I only have a problem with that the name "modify_health" is starting to get kind of inacurate, as I feel like it's doing more than that, which could lead to confusion down the line. I think such an important function as the main damage function should have a more intuitive name, though this might be outside of this issue's scope.

The function itself might be doing a bit too much and might get overly nested, which can make it hard to update later. So I would sugest extracting some of the functionality into helper functions, kind of like what `_set_health` is right now.

otherwise I think the solution is good.

**Comment by Turtyo** [comment]
I agree with the name not being great. Changing a name is not as bad as changing the signature, as a simple find and replace can do the trick for all the files.
How about `manage_player_health_and_block` ?

Which parts would you like to extract into functions ? Usually extracting into functions is good for two cases:
- re-use
- clarity
So the part you want to put into functions should either be re-used, or be unclear on itself (with exceptions as always, like `reset_block` for example)

**Comment by JonaLam** [comment]
`manage_player_health_and_block` is better, but it kind of implies that you can add block with the function, but my only alternative is `take_damage` which might be to vauge.

my comment about extractions were mostly for clarity in the functions.  It wil make the function easier to read for when we might wanna change it. Also `modify_health`, is currently responsible for both taking damage, as well as healing, which might be a problem if we want to add for exsample damage animations and stuff like that.Though these things could easily be changed when needed, so it's not too big a deal right now.

**Comment by Turtyo** [comment]
`modify_health` has a boolean to say if it's healing or damage
There was a reason for which we didn't split into `take_damage` and `heal`, which had to do with stats.

I started typing an explication but I realize it might be better to actually split between a `damage` and a `heal` function. We'll need to be very vigilant regarding bugs and checks though, because I think there was a bug with negative amount of damage healing you. Changing `modify_health` into two functions mean that almost all effects will change; if you feel like doing it, it's probably a good think

So:
- `take_damage_block_and_health` which reduces the block then the heath
- `health_damage`
- `heal`
- `block_damage` (returns the overflow amount of damage to the health as I explained previously)
- `get_block`
